crates/engine/tree/src/engine.rs

```diff
@@ -0,0 +1,173 @@
+//! An engine API handler for the chain.
+
+use crate::{
+    chain::{ChainHandler, FromOrchestrator, HandlerEvent, OrchestratorState},
+    tree::EngineApiTreeHandler,
+};
+use reth_beacon_consensus::BeaconEngineMessage;
+use reth_primitives::{SealedBlockWithSenders, B256};
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Advances the chain based on incoming requests.
+///
+/// This is a general purpose request handler with network access.
+/// This type listens for incoming messages and processes them via the configured request handler.
+///
+/// ## Overview
+///
+/// This type is an orchestrator for incoming messages and responsible for delegating requests
+/// received from the CL to the handler.
+///
+/// It is responsible for handling the following:
+/// - Downloading blocks on demand from the network if requested by the [`EngineApiRequestHandler`].
+///
+/// The core logic is part of the [EngineRequestHandler], which is responsible for processing the
+/// incoming requests.
+pub struct EngineHandler<T>
+where
+    T: EngineRequestHandler,
+{
+    /// Processes requests.
+    ///
+    /// This type is responsible for processing incoming requests.
+    handler: T,
+    /// Receiver for incoming requests that need to be processed.
+    // TODO add stream type for T::Request,
+    incoming_requests: (),
+    /// Access to the network sync to download blocks on demand.
+    network_sync: (),
+    /// Requests that are buffered and need to be processed.
+    buffered_events: VecDeque<()>,
+}
+
+impl<T> ChainHandler for EngineHandler<T>
+where
+    T: EngineRequestHandler,
+{
+    fn on_event(&mut self, event: FromOrchestrator) {
+        todo!()
+    }
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<HandlerEvent> {
+        todo!()
+    }
+}
+
+/// A type that processes incoming requests (e.g. requests from the consensus layer, engine API)
+pub trait EngineRequestHandler: Send + Sync {
+    /// The request type this handler can process.
+    type Request;
+
+    /// Informs the handler about an event from the [`EngineHandler`].
+    fn on_event(&mut self, event: FromEngine<Self::Request>);
+
+    /// Advances the handler.
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<RequestHandlerEvent>;
+}
+
+/// An [`EngineRequestHandler`] that processes engine API requests.
+///
+/// This type is responsible for advancing the chain during live sync (following the tip of the
+/// chain).
+///
+/// It advances the chain based on received engine API requests:
+///
+/// - `on_new_payload`: Executes the payload and inserts it into the tree. These are allowed to be
+///   processed concurrently.
+/// - `on_forkchoice_updated`: Updates the fork choice based on the new head. These require write
+///   access to the database and are skipped if the handler can't acquire exclusive access to the
+///   database.
```

Comment:
Could on forkchoice updated buffer up the data instead and have a way to configure when the write to the db is done?

Reply:
buffering should be done wrt to time out https://github.com/paradigmxyz/reth/pull/8251#issuecomment-2126280795. we should use a data structure that checks if existing entries are expired on insert
```rust
fn buffer_fcu(&mut self, msg: ForkChoiceUpdate) {
    let mut iter = self.iter_mut().rev();
    let mut entry = iter.next();
    let mut expired_count = 0;
    while entry.is_some() && entry.unwrap().insert_time.elapsed() >= Duration::from_secs(8) {
        expired_count += 1;
    }
    self.split_off(self.len() - expired_count)
    
    self.push(msg)
}
```

Reply:
ref https://github.com/paradigmxyz/reth/issues/8579